[GEN][ZH] Fix undefined behavior in ThingTemplate::isEquivalentTo

Caball009 opened the pull request:
Another case of checking 'this' pointer in a member function like https://github.com/TheSuperHackers/GeneralsGameCode/pull/870 & https://github.com/TheSuperHackers/GeneralsGameCode/pull/878

https://github.com/TheSuperHackers/GeneralsGameCode/blob/15a53d5c43f643a9b26a97a57d84062deeb4cf5c/GeneralsMD/Code/GameEngine/Source/Common/Thing/ThingTemplate.cpp#L1479-L1484

This pr fixes the undefined behavior in `ThingTemplate::isEquivalentTo` by making it a static function.

If the changes are considered acceptable, then it needs to be replicated in Generals. As far as I'm concerned that can be done in this pr as well.

Caball009 commented on GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Update/ProductionUpdate.cpp:
```diff
@@ -443,7 +443,7 @@ Bool ProductionUpdate::queueCreateUnit( const ThingTemplate *unitType, Productio
 	for( std::vector<QuantityModifier>::const_iterator it = data->m_quantityModifiers.begin(); it != data->m_quantityModifiers.end(); ++it )
   {
 		const ThingTemplate* productionTemplate = TheThingFactory->findTemplate( it->m_templateName );
-		if( productionTemplate && productionTemplate->isEquivalentTo( unitType ) )
```
productionTemplate is checked in isEquivalentTo.

Mauller commented:
The better solution here might be to get rid of the equivalent to function and Instead implement an overload for the comparison operator.

Then do ```(object->getTemplate() == otherTemplate) ```

xezon commented on GeneralsMD/Code/GameEngine/Source/Common/Thing/ThingTemplate.cpp:
```diff
@@ -1476,43 +1476,43 @@ UnsignedInt ThingTemplate::getMaxSimultaneousOfType() const
+Bool ThingTemplate::isEquivalentTo(const ThingTemplate* lhs, const ThingTemplate* rhs)
 {
 	// sanity
-	if (!(this && tt)) 
```
There is a DEBUG_CRASH. It is a bug and we see it.

xezon replied:
I looked at this wrong. I thought `isEquivalentTo` is part of `Thing` and not `ThingTemplate`. There are only 28 callers to `isEquivalentTo`, so I think we can remove the this!=NULL test and fix the callers where so necessary.

Caball009 replied:
Some calls are like this, however, so it might work better to move the check to function:
https://github.com/TheSuperHackers/GeneralsGameCode/blob/9c20c27be58463a755b8c67a50f1fdf146a33b68/GeneralsMD/Code/GameEngine/Source/GameLogic/AI/AIPlayer.cpp#L1072-L1073

Caball009 replied:
> I looked at this wrong. I thought `isEquivalentTo` is part of `Thing` and not `ThingTemplate`. There are only 28 callers to `isEquivalentTo`, so I think we can remove the this!=NULL test and fix the callers where so necessary.

It's been a little while since I last looked at this. Which of those 28 (excluding `unit`) would warrant a check at the call site?

xezon replied:
Any ones that are unsafe. This for example:

```cpp
Bool RiderChangeContain::isValidContainerFor(const Object* rider, Bool checkCapacity) const
{
...
  const ThingTemplate *thing = TheThingFactory->findTemplate( data->m_riders[ i ].m_templateName );
  if( thing->isEquivalentTo( rider->getTemplate() ) ) // <--- thing could be NULL and must be tested
```

Caball009 commented:
> The better solution here might be to get rid of the equivalent to function and Instead implement an overload for the comparison operator within thingtemplate.
> 
> Then do `(object->getTemplate() == otherTemplate) `

I like that approach.

Caball009 commented:
Updated and replicated in Generals. 

I made the assumption that the `ThingTemplate` cannot be NULL for `Object`s and `PartitionFilterThing`s. Afaik, objects are only created in `ThingFactory::newObject`, which checks if the template is `NULL`. I checked and all calls to `PartitionFilterThing`s constructor are behind a template != `NULL` check. I added an assertion to the constructor for good measure.

xezon reviewed (approved):
Looks very safe.
